source/extensions/http/cache/file_system_http_cache/cache_file_fixed_block.cc

```diff
@@ -38,8 +38,66 @@ CacheFileFixedBlock::CacheFileFixedBlock() {
 }
 
 void CacheFileFixedBlock::populateFromStringView(absl::string_view s) {
-  ASSERT(s.size() == size());
-  safeMemcpyUnsafeSrc(&contents_, s.begin());
+  ASSERT(s.size() == size() && size() == 24);
+  // Bytewise copy the values from the string_view into the fields.
+  // This ensures consistent endianness, and is the only safe way to copy values
+  // from a byte array into value types.
+  // chars must be cast to uint8_t first so that e.g. -1 becomes 0xff, before casting to uint64_t
+  // - otherwise -1 becomes 0xffffffffffffffff.
+  file_id_ = static_cast<uint32_t>(static_cast<uint8_t>(s[0])) << 24 |
```

Comment:
this was my punishment for being too picky :)

I think the previous state was obviously better from a readability perspective.

Reply:
Also I was thinking you could still read into the struct but just correct the endian-ness (if needed) when you do that.

For writing I guess you could either copy the FileBlock out into a temp and twiddle the bits (if you need to) before appending to the Envoy buffer. Compared with the cost of fileI/O that extra buffer copy seems pretty miniscule.

One more thing: do you want to make size be 64 bits?

Reply:
Well, the struct was relatively problematic anyway (it's *borderline* undefined behavior to union bytes over a bunch of ints, especially with the dangers of packing). Having made this change I'm reluctant to change it back. I have made the serialize-on-demand more readable with some inline functions for it. I think the case for making these functions 'common' is weaker than it was for endian-swap functions, but I'm not sure it's something nobody else would want (I wasn't able to find any existing impl this time though!)

I'm more comfortable with body size as 64 bits because, yeah, maybe someone *really does* want to put a full length HD movie into a cache! I don't want to be accidentally creating the next FAT32 situation. (I'm even a little nervous about 32-bit headers and trailers sizes - sure *today* they're generally smaller than 1K, but who knows what madness the future holds!)

Reply:
(There is [this](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/grpc/codec.cc?L19:43) suggesting maybe there are other uses for serializeUint32, but that's the only existing one, in only one direction, and it's not going onto chars, so I think it's more likely that each use-case is sufficiently unique that it's not worth putting in common.)

Reply:
I am pretty slammed today (prepping for an exec review), so I may not have time for another round, but please explain where the UB is?

I think the main thing in my mind is that to avoid  a spurious uninitialized read you should memset the structure before populating it, depending on how you populate it.

But apart from that I don't see the UB.

Reply:
I'm trying to understand the practical implication of what you are saying. The downside of using the struct is that there might be some platform in the future where two uint32_t are each padded to 8-byte boundaries, and you want to be able to share cache files between that platform and an arm or x86. Is that right?

I question the premise of ever sharing these files cross-platform in the first place, but if it's not too expensive to support I can't complain if it's possible. E.g. storing the data as little-endian and calling an existing conversion to the native type (which would be a no-op on x86 and arm) seems pretty reasonable :)

But this approach you have here is a whole lot of byte-hacking code that needs to be reviewed, not just by me, but by future people concerned about potential security issues. So I am not sure it's worth it to allow file-sharing with future architectures.

Like I said you could easily write a platform signature to the top of the structure and reject reading files that don't match the current architecture, saving a lot of cognitive load.